### PR TITLE
Nonuniform sampling support

### PR DESCRIPTION
**Pull request by filssavi**
solution to issue #174

### CODE REVIEW

**Comment by filssavi** [comment]
Upon further testing there are few bugs still left to iron out, moreover as the distance from the bound increases a spike the edge of the limited sample's distibution is introduced, I have to sort this issue prior to merging

So I will close this pull request and make a new one once this issues are fixed

**Comment by filssavi** [comment]
the last bugs have been ironed out and the problems at the bounds have been fixed by proper rescaling prior to the distribution change

**Comment by jdherman** [comment]
Hi @filssavi thanks for working on this. Nonuniform sampling will be a nice addition to the other methods.

I didn't think it would need to be so complicated, though. In the `saltelli` sampling method (the one that existed before), the code is only this:
```python
    if not problem.get('dists'):
        # scaling values out of 0-1 range with uniform distributions
        scale_samples(saltelli_sequence, problem['bounds'])
        return saltelli_sequence
    else:
        # scaling values to other distributions based on inverse CDFs
        scaled_saltelli = nonuniform_scale_samples(saltelli_sequence, problem['bounds'], problem['dists'])
        return scaled_saltelli
```

I thought we could just add a similar if-then block to the other methods. In this PR it seems that things are implemented differently across methods, and I also don't understand the need for any bounds checking or cropping the range of samples.

Part of the confusion could be the way that `nonuniform_scale_samples` works, because the parameters that we call "lower and upper bounds" (for uniform distributions) are used differently for the other distributions:
```python
def nonuniform_scale_samples(params, bounds, dists):
    """Rescale samples in 0-to-1 range to other distributions

    Arguments
    ---------
    problem : dict
        problem definition including bounds
    params : numpy.ndarray
        numpy array of dimensions num_params-by-N,
        where N is the number of samples
    dists : list
        list of distributions, one for each parameter
            unif: uniform with lower and upper bounds
            triang: triangular with width (scale) and location of peak
                    location of peak is in percentage of width
                    lower bound assumed to be zero
            norm: normal distribution with mean and standard deviation
            lognorm: lognormal with ln-space mean and standard deviation
    """
```
The samples from the nonuniform distributions do not need to fall between any bounds. The [0,1] samples are put into the inverse CDF to generate samples from the chosen distribution.

So, the implementation can be simpler and cleaner across methods. I appreciate the time investment and will keep the PR open if you want to keep working on it.

Thanks,
Jon

**Comment by filssavi** [comment]
The first test I did was just added the if-then-else block from saltelli sampling to the method of morris one, however (probably due to differences in the sample generating algorithms infinities started to come out

The core problem is that for infinite distributions (normal and lognormal in this case) the following are true:

codecogseqn
codecogseqn 1

so if 0 or 1 is present in the original (unscaled and unshifted) samples list, when resampling with  `nonuniform_scale_samples` -inf and +inf will be output, this while technically correct is unhelpful, since:

- most models are not built to take into account infinite inputs so they will most definitely crash or refuse to work entirely

- when using sensitivity analysis to asses the validity of model of some phenomenon happening in the real world, the values that a parameter, might follow a normal/lognormal distribution, but some nonlinear effect will limit the range of possible values, an example (from wikipedia) is  blood pressure of adult humans, however at some point blood vessels will start to rupture, so the tails of the distributions will not go up to infinity


For this reason, the values of the original samples (before any rescaling), so a reasonable assumption is that for a normal distribution a reasonable default could be to consider the six sigma range around the mean value.

However since it comes essentially for free I thought to offer users of the library the possibility of overriding this default and manually choosing the range of the normal distribution form which to sample,

in the case of the saltelli sampling method (with which i'm not too familiar so what follows it's pure speculation), I think this problems haven't shown up because the original samples are already chosen in a way that excludes the boundary values

File: SALib/sample/ff.py [hunk 1]
```diff
@@ -76,30 +76,24 @@ def generate_contrast(problem):
 
 def sample(problem):
     """Generates model inputs using a fractional factorial sample
-
```
Comment (willu47) [inline SALib/sample/ff.py]: These blank lines should be retained in the docstring to comply with numpydoc style.

File: SALib/sample/latin.py [hunk 2]
```diff
@@ -39,8 +39,23 @@ def sample(problem, N):
         for j in range(N):
             result[j, i] = temp[j]
 
-    scale_samples(result, problem['bounds'])
```
Comment (willu47) [inline SALib/sample/latin.py]: Could all this functionality be pushed into the `scale_samples` function if the problem dict was passed to it?  This way, there would be no duplication of the code across the different methods?

File: SALib/util/__init__.py [hunk 3]
```diff
@@ -280,3 +280,70 @@ def _wrapper(*args, **kwargs):
...
+            bounds.append([lower_bound,upper_bound])
+        else:
+            bounds.append([0,1])
+
+    limited_sample = samples
+    scale_samples(samples,bounds)
+
+    return limited_sample
+
+def checkBounds(problem):
```
Comment (willu47) [inline SALib/util/__init__.py]: To comply with python style guides, camel case should not be used please use lower case and underscores for function names.

File: SALib/util/__init__.py [hunk 4]
```diff
@@ -280,3 +280,70 @@ def _wrapper(*args, **kwargs):
...
+    """
+    if not problem.get('dists_upper_bound'):
+        upper_bound = 0.999999998026825
+    else:
+        upper_bound = problem.get('dists_upper_bound')
+        if not 0 < upper_bound < 1:
+            raise ValueError("Nonuniform distribution value range invalid")
+
+    if not problem.get('dists_lower_bound'):
+        lower_bound = 1 - 0.999999998026825
```
Comment (willu47) [inline SALib/util/__init__.py]: `lower_bound = 1 - upper_bound` would be neater

**Comment by filssavi** [comment]
I implemented most of the required changes (tests are still missing, i'll add them when I have some time)

the only issue remaining is with the fast method, somehow the change makes the regression test fail (the analisys results are different), I checked and superficially there seems to be nothing wrong with the sampling part of the code

since I don't really know how fast inner working and all other tests are passing I suspect the sampling part is working fine and I somehow triggered a bug in the fast analize implementation, can someone confirm/deny this

thanks

**Comment by willu47** [comment]
@filssavi - I'm now back in my office and can take another look at these failing tests and get this pull request merged soon.

**Comment by willu47** [comment]
Closing this branch in preference to the rebased (and renamed from filssavi/master) #192.

**Pull request closed without merging**